posts/MigueTripi/01.isContractIssue.md

```diff
@@ -0,0 +1,56 @@
+---
+title: 'Using OpenZepellin standard IsContract() could be unsafe'
+published: false
+description: We are at risk if we do not use this consciously
+tags: 'Solidity, EVM, Etherium, hack'
+cover_image: ./assets/01.cover.png
+id: 
+date: '2023-01-02T12:12:12Z'
+---
+
+
+## Context
+
+Most of developers follow industry's standard to avoid re-create the wheel and fall into common issues. Solidity world is not an exception of it.
+One of the most popular standard is OpenZepellin library which provide us several usefull implementations such as ERC20 token, ERC721 token, Governor, etc.
+
+## Issue
+
+But well not everything in the standard must be used without taking into account the risks we could face in the future. As fas as we know, in public blockchains like Etherium this potentially means drained funds from hackers.
+Let my bring a method which was discussed several times last year.
+```Address.isContract()```
+
+## Explanation
+
+This routine returns true in case the provided address contains some code which makes sense. Well, actually there are some special situations where the condition gets broken:
+
+- During contract creation the code of the contract is equals to zero. This breaks the logic due to if you could call this method from the contract's constructor the method will return false.
+- If you send an address where a new contract will be located in the fuutre.
+- If you send an address where were located and deleted some contract.
+
+## Risks
+
+There is at least one common scenario where your protocol could get into troubles:
+In case you just call some address assuming it is not a contract you could open a window for reentrancy attack.
+
+So, how to correctly validate if an address is a contract?
+
+## Proposed WA
+
+Actually it is a sort of limitation placed in the EVM. I guess it would have several ways to solve but it is not known when a solution will take place.
+
+Sometime ago I faced a misuse of this function during an audition in Code4rena platform. For this issue I proposed a sort of public whitelisting of the address where the final user could:
+Add the address as a whilisted EOA with the current blocknumber.
+Before doing the stuff we have to:
+
+- Validate the whitelisting.
+- Validate that currentBlock is higher than the whitelisting block number.
+- Finally, validate Address.isContract() method from OZ standard.
+
+These steps are not the most user friendly flow but it works if you want to avoid call some contract.
+
+## Next steps
+
+As I mentioned at the beggining, this issue was discussed several times during last year. Currenly there is an open PR in OZ repository to delete `isContract()` method in OZ version 5.0. But becareful because this method will be marked as deprecated (but able to use) on previous versions.
```

Comment:
Typo en **beginning**
Typo en **be careful**

Reply:
Done!